tests/src/test/scala/whisk/core/containerpool/docker/test/AkkaContainerClientTests.scala

```diff
@@ -0,0 +1,213 @@
+    val result = Await.result(connection.post("/init", JsObject.empty, retry = true), 10.seconds)
+    val end = Instant.now()
+    val waited = end.toEpochMilli - start.toEpochMilli
+    result should be('left)
+    result.left.get shouldBe a[Timeout]
+    result.left.get.asInstanceOf[Timeout].t shouldBe a[TimeoutException]
+
+    waited should be > timeout.toMillis
+    waited should be < (timeout * 2).toMillis
+  }
+
+  it should "retry till success within timeout limit" in {
+    val timeout = 5.seconds
+    val retryInterval = 500.milliseconds
+    val connection =
+      new AkkaContainerClient(httpHost.getHostName, httpHost.getPort, timeout, 1.B, 100, retryInterval)
+    val start = Instant.now()
+    testConnectionFailCount = 5
+    testResponse = ""
+    val result = Await.result(connection.post("/init", JsObject.empty, retry = true), 10.seconds)
+    val end = Instant.now()
+    val waited = end.toEpochMilli - start.toEpochMilli
+    result shouldBe Right {
+      ContainerResponse(true, "", None)
+    }
+
+    waited should be > (testConnectionFailCount * retryInterval).toMillis
+    waited should be < timeout.toMillis
+  }
+
+  it should "not truncate responses within limit" in {
+    val timeout = 1.minute.toMillis
+    val connection = new AkkaContainerClient(httpHost.getHostName, httpHost.getPort, timeout.millis, 50.B, 100)
+    Seq(true, false).foreach { code =>
+      Seq(null, "", "abc", """{"a":"B"}""", """["a", "b"]""").foreach { r =>
+        testStatusCode = if (code) 200 else 500
+        testResponse = r
+        val result = Await.result(connection.post("/init", JsObject.empty, retry = true), 10.seconds)
+        result shouldBe Right {
+          ContainerResponse(okStatus = code, if (r != null) r else "", None)
+        }
+      }
+    }
+  }
+
+  it should "truncate responses that exceed limit" in {
+    val timeout = 1.minute.toMillis
+    val limit = 1.B
+    val connection = new AkkaContainerClient(httpHost.getHostName, httpHost.getPort, timeout.millis, limit, 100)
+    Seq(true, false).foreach { code =>
+      Seq("abc", """{"a":"B"}""", """["a", "b"]""").foreach { r =>
+        testStatusCode = if (code) 200 else 500
+        testResponse = r
+        val result = Await.result(connection.post("/init", JsObject.empty, retry = true), 10.seconds)
+        result shouldBe Right {
+          ContainerResponse(okStatus = code, r.take(limit.toBytes.toInt), Some((r.length.B, limit)))
+        }
+      }
+    }
+  }
+
+  it should "truncate large responses that exceed limit" in {
+    val timeout = 1.minute.toMillis
+    //use a limit large enough to not fit into a single ByteString as response entity is parsed into multiple ByteStrings
+    //seems like this varies, but often is ~64k or ~128k
+    val limit = 300.KB
+    val connection = new AkkaContainerClient(httpHost.getHostName, httpHost.getPort, timeout.millis, limit, 100)
+    Seq(true, false).foreach { code =>
+      Seq("0123456789" * 100000).foreach { r =>
```

Comment:
Why even use `Seq.foreach` if you only pass in one value?

Please add a comment on what that value is supposed to be, i.e.

```scala
// Generate a response that's 1MB
val response = "0123456789" * 1024 * 1024
```

To make the numbers less magic.